zipline/pipeline/loaders/blaze/earnings.py

```diff
@@ -106,33 +79,15 @@ def __init__(self,
         self._dataset = dataset
 
     def load_adjusted_array(self, columns, dates, assets, mask):
-        expr = self._expr
-        filtered = expr[expr[TS_FIELD_NAME] <= dates[0]]
-        lower = odo(
-            bz.by(
-                filtered[SID_FIELD_NAME],
-                timestamp=filtered[TS_FIELD_NAME].max(),
-            ).timestamp.min(),
-            pd.Timestamp,
-            **self._odo_kwargs
-        )
-        if pd.isnull(lower):
-            # If there is no lower date, just query for data in the date
-            # range. It must all be null anyways.
-            lower = dates[0]
-
-        raw = odo(
-            expr[
-                (expr[TS_FIELD_NAME] >= lower) &
-                (expr[TS_FIELD_NAME] <= dates[-1])
-            ],
-            pd.DataFrame,
-            **self._odo_kwargs
+        raw = ffill_query_in_range(
+            self._expr,
+            dates[0],
+            dates[-1],
+            self._odo_kwargs,
         )
-
         sids = raw.loc[:, SID_FIELD_NAME]
         raw.drop(
-            sids[~(sids.isin(assets) | sids.notnull())].index,
```

Comment:
This is just prefiltering before doing the work in the dataframe loader. The final output will be the same but should be faster.